src/shogun/mathematics/graph/Graph.cpp

```diff
@@ -22,12 +22,15 @@ Graph::Graph(
 
 void Graph::build()
 {
-	auto* env = ShogunEnv::instance();
-	build(env->graph_backend());
+	auto* env = GraphEnv::instance();
+	build(env->get_backend());
 }
 
 void Graph::build(GRAPH_BACKEND backend)
 {
+	auto* env = GraphEnv::instance();
+	std::lock_guard<std::mutex> lock(env->m_env_mutex);
```

Comment:
@vigsterkr does this make sense? Basically lock the mutex here so that you can't change the backend in another thread whilst the graph is building

Reply:
yep!

Reply:
the only thing is that in the code above `build(env->get_backend());` in `Graph::build()` is not locked, so there is a chance that the backend changes there.. right?